src/python/pants/backend/python/subsystems/pex_build_util.py

```diff
@@ -148,16 +149,33 @@ def add_requirement_libs_from(self, req_libs, platforms=None):
     reqs = [req for req_lib in req_libs for req in req_lib.requirements]
     self.add_resolved_requirements(reqs, platforms=platforms)
 
-  def add_resolved_requirements(self, reqs, platforms=None):
-    """Multi-platform dependency resolution for PEX files.
+  class SingleDistExtractionError(Exception): pass
 
-    :param builder: Dump the requirements into this builder.
-    :param interpreter: The :class:`PythonInterpreter` to resolve requirements for.
-    :param reqs: A list of :class:`PythonRequirement` to resolve.
-    :param log: Use this logger.
-    :param platforms: A list of :class:`Platform`s to resolve requirements for.
-                      Defaults to the platforms specified by PythonSetup.
+  def extract_single_dist_for_current_platform(self, reqs, dist_key):
+    """Resolve a specific distribution from a set of requirements matching the current platform.
+
+    :param list reqs: A list of :class:`PythonRequirement` to resolve.
+    :param str dist_key: The value of `distribution.key` to match for a `distribution` from the
+                         resolved requirements.
+    :return: The single :class:`pkg_resources.Distribution` matching `dist_key`.
+    :raises: :class:`self.SingleDistExtractionError` if no dists or multiple dists matched the given
+             `dist_key`.
     """
+    distributions = self._resolve_distributions_by_platform(reqs, platforms=['current'])
+    try:
+      matched_dist = assert_single_element(list(
+        dist
+        for _, dists in distributions.items()
+        for dist in dists
+        if dist.key == dist_key
+      ))
+    except (StopIteration, ValueError) as e:
+      raise self.SingleDistExtractionError(
+        "Exactly one dist was expected to match name {} in requirements {}: {}"
+        .format(dist_key, reqs, e))
```

Comment:
It'd be nice to have a test that exercises this case, if there isn't already.

Reply:
Currently there is no testing except the testprojects integration test :) Will look into that now.

Reply:
Done! This testing exposed the fact that we don't always want to descend into the `<name>-<version>.data/` directory (although we probably do most of the time) -- so also specialized that to occur if the new `within_data_subdir=` kwarg is set. So this testing was extremely useful although there were multiple TODOs left over.